riscv/mmu.h

```diff
@@ -290,8 +290,10 @@ class mmu_t
     if (matched_trigger)
       throw *matched_trigger;
 
+    reg_t vpn = addr >> PGSHIFT;
     auto access_info = generate_access_info(addr, FETCH, {false, false, false});
-    check_triggers(triggers::OPERATION_EXECUTE, addr, access_info.effective_virt);
+    if (unlikely(tlb_insn_tag[vpn % TLB_ENTRIES] != vpn))
+      check_triggers(triggers::OPERATION_EXECUTE, addr, access_info.effective_virt);
```

Comment:
I believe this will spuriously invoke `check_triggers` on ITLB misses.  Note that fetches from MMIO devices accesses always result in ITLB misses, so this means `check_triggers` will be invoked on every MMIO fetch.  Maybe this is functionally correct and is only a perf issue, but it seems undesirable.

What's wrong with my original suggestion of moving this check back into `fetch_slow_path`?  By construction, we should not need to check the triggers on the fetch fast path...

Reply:
The current implementation splits instruction fetching into multiple 2-byte operations, which call ```fetch_slow_path``` separately. This results in that a 4-byte instruction 0xe9830313 matches data triggers with tdata2={0xe983, 0x0313} instead of one with tdata2=0xe9830313 incorrectly. Additionally, a 4-byte instruction with PC at 0x4000 matches address triggers with tdata2={0x4000, 0x4002}, which is unexpected from a user perspective.

The behavior of the data trigger in the current implementation does not conform to the specification. It is weird to me that the address trigger matches PCs of 2-byte granularity. Clarify whether the specification allows the address trigger to match an arbitrary subset of the accessing addresses: https://github.com/riscv/riscv-debug-spec/issues/877.